src/sqlfluff/core/rules/analysis/select_crawler.py

```diff
@@ -273,10 +274,14 @@ def finish_segment():
                                 append_query(query)
                     else:
                         # We're processing a "with" statement.
-                        if cte_name:
+                        if cte_name_segment:
                             # If we have a CTE name, this is the Query for that
                             # name.
-                            query = self.query_class(QueryType.Simple, dialect)
+                            query = self.query_class(
+                                QueryType.Simple,
+                                dialect,
+                                cte_name_segment=cte_name_segment,
+                            )
```

Comment:
Now  the `Query` objects returned  by `SelectCrawler` can tell you the segment that each CTE's name.